### PR TITLE
enhancement(nats source): comply with component spec

### LINKED ISSUE #10072: Instrument the `nats` source with event processing metrics
Per #9326

### PR DESCRIPTION
**Pull request by jdrouet**
Adds `BytesReceived`, `StreamClosedError` and update `NatsEventsReceived`.
Also converts `out.send` to `out.send_all`.

closes #10072

Signed-off-by: Jérémie Drouet <jeremie.drouet@datadoghq.com>

### CODE REVIEW

**Review by 001wwang — APPROVED** [review]
Nice, just one very minor suggestion.

File: src/sources/nats.rs [hunk 1]
```diff
@@ -132,29 +132,30 @@ async fn nats_source(
...
         let mut stream = FramedRead::new(msg.data.as_ref(), decoder.clone());
         while let Some(next) = stream.next().await {
             match next {
                 Ok((events, byte_size)) => {
-                    emit!(&NatsEventsReceived {
-                        byte_size,
-                        count: events.len()
-                    });
+                    let count = events.len();
+                    emit!(&NatsEventsReceived { byte_size, count });
```
Comment (jszwedko) [inline src/sources/nats.rs:144]: I don't think the `byte_size` here is right. The `byte_size` returned by the decoder is actually the number of bytes from the frame, so basically the same as received bytes above that is published as `component_received_bytes_total`:

https://github.com/vectordotdev/vector/blob/86bfec8d70cb30158964a0208d37b30723be3e8f/src/codecs/decoding.rs#L191-L216

Instead, `component_received_event_bytes_total` should reflect the total deserialized event bytes in-memory. That is, you should sum up `ByteSizeOf::size_of` of the events here.

**Review by jszwedko — CHANGES_REQUESTED** [review]
One question about the accounting of `component_received_event_bytes_total` here.

**Review by jszwedko — APPROVED** [review]
Thanks for the updates!

**Pull request merged**